File: main/crc.c
#include "crc.h"
#include "esp_log.h"

/* FUNCTION PROTOTYPE */
static uint32_t reflect(uint32_t data, int bits);
static uint32_t crc32(const uint8_t *data, uint32_t length);

/**
 * @brief
 * @param
 * @param
 * @return
 * */
static uint32_t reflect(uint32_t data, int bits) {
    uint32_t reflection = 0;
    for (int i = 0; i < bits; i++) {
        if (data & (1 << i)) {
            reflection |= (1 << (bits - 1 - i));
        }
    }
    return reflection;
}

static uint32_t crc32(const uint8_t *data, uint32_t length) {
    uint32_t crc = 0xFFFFFFFF;
    uint32_t poly = 0x04C11DB7;
    
    for (uint32_t i = 0; i < length; i++) {
        ESP_LOGI("BYTE", "0x%2X", data[i]);
        uint8_t byte = reflect(data[i], 8);
        crc ^= ((uint32_t)byte) << 24;

        for (int j = 0; j < 8; j++) {
            if (crc & 0x80000000)
                crc = (crc << 1) ^ poly;
            else
                crc <<= 1;
        }
        crc &= 0xFFFFFFFF;
    }

    crc = reflect(crc, 32);
    ESP_LOGI("CRC", "0x%2X", (int)(crc^0xFFFFFFFF));
    return crc ^ 0xFFFFFFFF;
}

bool validate_crc(const uint8_t* data, uint32_t length, uint32_t crc_in_packet){
    uint32_t crc_calc = crc32(data, length);
    ESP_LOGI("CRC CALCULATED", "0x%X", (int)crc_calc);

    if (crc_calc == crc_in_packet) {
        return true;
    }
    else {
        return false;
    }
}

File: python_testing/crc.py
def reflect(value, bits):
    result = 0
    for i in range(bits):
        if value & (1 << i):
            result |= 1 << (bits - 1 - i)
    return result

def crc32_iso(data: bytes) -> int:
    poly = 0x04C11DB7
    crc = 0xFFFFFFFF

    for byte in data:
        byte = reflect(byte, 8)
        crc ^= byte << 24

        for _ in range(8):
            if crc & 0x80000000:
                crc = (crc << 1) ^ poly
            else:
                crc <<= 1
            crc &= 0xFFFFFFFF  # Ensure 32-bit range

    crc = reflect(crc, 32)
    return crc ^ 0xFFFFFFFF

# Test
# data = bytes([0xA5, 0x50, 0x00])
value = 0xA55000
byte_array = value.to_bytes(3, byteorder='big')  # 3 bytes for 24 bits
print(f"CRC32 (ISO): 0x{crc32_iso(byte_array):08X}")

# value = 0xA55000
# byte_array = value.to_bytes(3, byteorder='big')  # 3 bytes for 24 bits
# print(byte_array)           # Output: b'\xa5P\x00'
# a = crc32(byte_array)
# crc = zlib.crc32(byte_array) & 0xFFFFFFFF
# print(f"CRC32 (ISO): 0x{crc:08X}")
# print(hex(a))